File: backend/src/main/java/com/nottingham/mynottingham/backend/entity/Errand.java
package com.nottingham.mynottingham.backend.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

@Data
@Entity
@Table(name = "errands")
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = true)
public class Errand extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "requester_id", nullable = false)
    @JsonIgnoreProperties({"password"})
    private User requester;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "provider_id")
    @JsonIgnoreProperties({"password"})
    private User provider;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(length = 1000)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 50)
    private ErrandType type;

    @Column(nullable = false, length = 200)
    private String location;

    @Column(length = 100)
    private String deadline;

    @Column(nullable = false)
    private Double reward = 0.0;

    @Column(length = 500)
    private String additionalNotes;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ErrandStatus status = ErrandStatus.PENDING;

    private String imageUrl;

    public enum ErrandType {
        SHOPPING, PICKUP, FOOD_DELIVERY, OTHER
    }

    public enum ErrandStatus {
        PENDING, IN_PROGRESS, COMPLETED, CANCELLED
    }
}

File: app/src/main/java/com/nottingham/mynottingham/data/model/Errand.kt
package com.nottingham.mynottingham.data.model

/**
 * Errand data model
 */
data class Errand(
    val id: String,
    val requesterId: String,
    val requesterName: String,
    val requesterAvatar: String? = null,
    val requesterRating: Double = 0.0,
    val requesterReviewCount: Int = 0,
    val providerId: String? = null,
    val providerName: String? = null,
    val title: String,
    val description: String,
    val type: ErrandType,
    val location: String,
    val deadline: String,
    val reward: Double,
    val additionalNotes: String? = null,
    val status: ErrandStatus,
    val imageUrl: String? = null,
    val createdAt: Long,
    val updatedAt: Long,
    val completedAt: Long? = null
)

enum class ErrandType(val displayName: String, val color: String) {
    SHOPPING("Shopping", "#FF6B9D"),
    PICKUP("Pickup", "#4ECDC4"),
    FOOD_DELIVERY("Food Delivery", "#FFD93D"),
    OTHER("Others", "#A8E6CF")
}

enum class ErrandStatus {
    PENDING,      // Waiting to be taken
    ACCEPTED,     // Accepted (courier accepted but not started delivery)
    DELIVERING,   // In delivery
    IN_PROGRESS,  // Compatible with old status, equivalent to ACCEPTED
    COMPLETED,    // Completed
    CANCELLED     // Cancelled
}

/**
 * User balance for Campus Errand
 */
data class ErrandBalance(
    val userId: String,
    val balance: Double,
    val totalEarned: Double,
    val totalSpent: Double
)
